xline/src/storage/db.rs

```diff
@@ -1,141 +1,84 @@
-use parking_lot::{RwLock, RwLockUpgradableReadGuard};
-use prost::Message;
+use std::sync::Arc;
 
-use super::{storage_api::StorageApi, ExecuteError, Revision};
-use crate::rpc::{KeyValue, PbLease};
+use engine::{engine_api::StorageEngine, Delete, Put, WriteOperation};
+
+use super::{storage_api::StorageApi, ExecuteError};
 
 /// Database to store revision to kv mapping
-#[derive(Debug)]
-pub(crate) struct DB<S>
-where
-    S: StorageApi,
-{
+#[derive(Debug, Clone)]
+pub struct DB<S: StorageEngine> {
     /// internal storage of `DB`
-    storage: RwLock<S>,
+    engine: Arc<S>,
 }
 
 impl<S> DB<S>
 where
-    S: StorageApi,
+    S: StorageEngine,
 {
     /// New `DB`
-    pub(crate) fn new(storage: S) -> Self {
+    #[allow(dead_code)]
+    #[inline]
+    #[must_use]
+    pub fn new(engine: S) -> Self {
         Self {
-            storage: RwLock::new(storage),
+            engine: Arc::new(engine),
         }
     }
+}
+
+impl<S> StorageApi for DB<S>
```

Comment:
The interfaces provided by `StorageApi` and `StorageEngine` are very similar, they both have the ability to read and write `Vec<u8>`, why can't we merge them

Reply:
That's because Xline Server is not the only one with persistence requirements. So has the Curp Server. Their `StorageApi` may be different in that `StorageApi` is closer to the business logic. To isolate change, I add a middle trait `StorageEngine`. If we need to replace our storage engine in the future, we just need to implement `StorageEngine` for the new storage engine. But without that, we need to implement `StorageApi` twice, one for the Curp Server while the other for the Xline Server.